authorize/evaluator/evaluator.go

```diff
@@ -142,61 +288,224 @@ func New(
 }
 
 type routeEvaluator struct {
-	id        uint64
-	evaluator *PolicyEvaluator
+	id               uint64
+	evaluator        *PolicyEvaluator
+	computedChecksum uint64
+}
+
+var (
+	workerPoolSize      = runtime.NumCPU() - 1
+	workerPoolTaskQueue = make(chan func(), (workerPoolSize+1)*2)
+)
+
+func init() {
+	for i := 0; i < workerPoolSize; i++ {
+		go worker()
+	}
+}
+
+func worker() {
+	for fn := range workerPoolTaskQueue {
+		fn()
+	}
+}
+
+const chunkSize = 64
+
+type workerContext struct {
+	context.Context
+	Cfg           *evaluatorConfig
+	Store         *store.Store
+	StatusBits    []uint64
+	StatusCounts  []uint8
+	TotalModified *atomic.Int32
+	Evaluators    []routeEvaluator
+	EvalCache     *EvaluatorCache
+	QueryCache    *QueryCache
+	Errs          *sync.Map
+}
+
+type partition struct{ Begin, End int }
+
+func computeChecksums(wctx *workerContext, part partition) {
+	defer rttrace.StartRegion(wctx, "worker-checksum").End()
+	for chunkIdx := part.Begin; chunkIdx < part.End; chunkIdx++ {
+		var chunkStatus uint64
+		off := chunkIdx * chunkSize
+		limit := min(chunkSize, len(wctx.Cfg.Policies)-int(off))
+		for i := 0; i < limit; i++ {
+			p := wctx.Cfg.Policies[off+i]
+			id, err := p.RouteID()
+			if err != nil {
+				wctx.Errs.Store(off+i, fmt.Errorf("authorize: error computing policy route id: %w", err))
+				continue
+			}
+			eval := &wctx.Evaluators[off+i]
+			eval.id = id
+			eval.computedChecksum = p.ChecksumWithID(id)
+			cached, ok := wctx.EvalCache.LookupEvaluator(id)
+			if !ok {
+				rttrace.Logf(wctx, "", "policy for route ID %d not found in cache", id)
+				chunkStatus |= 1 << i
+			} else if cached.policyChecksum != eval.computedChecksum {
+				rttrace.Logf(wctx, "", "policy for route ID %d changed", id)
+				chunkStatus |= 1 << i
+			}
+		}
+		wctx.StatusBits[chunkIdx] = chunkStatus
+		popcnt := bits.OnesCount64(chunkStatus)
+		rttrace.Logf(wctx, "", "chunk %d: %d/%d changed", chunkIdx, popcnt, limit)
+		wctx.StatusCounts[chunkIdx] = uint8(popcnt)
+		wctx.TotalModified.Add(int32(popcnt))
+	}
+}
```

Comment:
I'm sorry but I'm not really following what this code is doing. I didn't really understand the original code either.

Reply:
(I mean all this building, computing checksums, worker chunking, bit manipulation, etc)

I think there are too many things going on here put together and I can't pull them apart to understand it.

Reply:
Yeah sorry this definitely needs a big comment explaining what's going on :smile: